### PR TITLE
Allow reordering of sidebar pages

### PR DESCRIPTION
**Pull request by luisbettencourt**
## 🌟 What is the purpose of this PR?

This PR adds the functionality to reorder pages in the sidebar and saving that order in the database.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1200211978612931/1202668694585973/f) _(internal)_

## 🔍 What does this change?

- Added drag and drop functionality to the page tree view in the sidebar;
- Added a loading state to the page section in the sidebar, ~preventing the user from creating pages or reordering pages when changes are being saved~.
- Added an index to pages to save their order in the sidebar (using fractional indexing to minimize the number of saves to the db);
- Created the getIncomingLinks query to support the reorder pages request.

## ⚠️ Known issues

- Pages are now required to have an index, so pages that already existed before this PR won't have an index and will always be ordered last on its parent, until they are dragged into a new position so that an index is assigned to them.

- When reordering a page, an absolute index is being sent to the API, so that the new fractional index for that page is calculated when the request happens. Because of this, reordering a page requires fetching all pages in that group (all account pages if the page is at the root level or all sibling pages if the page belongs to a parent page). This can take longer depending on the number of pages the user has created in his account (I've seen this request take about 5 to 6 seconds on a page with many archived pages).

  An issue with this is that a user could reorder a number of pages one right after the other and create conflicts in the page tree. However, after some testing I found the current approach quite reliable and couldn't reproduce this (but please, feel free to test it and try to prove me wrong on this!).

  The obvious solution to this would be to have the fractional index be calculated in the frontend and sending it in the request, only requiring an update to the page being reordered. This would significantly speed up the request, the only issue being that we could have issues if the page tree being displayed in the frontend is for some reason out of sync with the database.

## 🐾 Next steps

- [x] Fix the following animation bug:

bug

- [x] Add back the expand/collapse animation on page items.

## ❓ How to test this?

Try to abuse the page tree by reordering and reparenting pages.

## 📹 Demo

demo

### CODE REVIEW

**Comment by thehabbos007** [comment]
This is really cool, Luís! 

Just wanted to write down the idea that I had when we talked offline, I'm not sure it would help much after looking at the code. But here is the general outline:

What I was thinking is that given the following page structure:
<details>
<summary>Screenshot</summary>

image


</details>

Which has the following ordering current (identation means a parent relation):

<details>
<summary>Current page properties</summary>

```typescript
[
  {
    index: "a0",
    title: "Untitled",
  },
  {
    index: "a1",
    title: "My awesome page",
  },
  {
    index: "a2",
    title: "Top page",
  },
      {
        index: "Zy",
        title: "Middle page",
      },
          {
            index: "a0",
            title: "Table Testing Page",
          },
      {
        index: "Zz",
        title: "End page",
      },
]
```

</details>

we could instead have the indexing happen across all pages, regardless of the the parent relations, such that we get the following indexing:

<details>
<summary>Proposed page properties</summary>

```typescript
[
  {
    index: "a0",
    title: "Untitled",
  },
  {
    index: "a1",
    title: "My awesome page",
  },
  {
    index: "a2",
    title: "Top page",
  },
      {
        index: "a3",
        title: "Middle page",
      },
          {
            index: "a4",
            title: "Table Testing Page",
          },
      {
        // still follows the rest of the indices, but just not continuation of "a"
        index: "Zz", 
        title: "End page",
      },
]
```

</details>

because the parent relation does not influence the order of the pages. 

If I were to move `"End page"` (index `"Zz"`) above `"Middle page"` (index `"a3"`), the index would change on `"End page"` but not necessarily remove its parent page relationship. 

I would imagine this would simplify some of the ordering logic a little, but from my local testing of the current code, I'm not seeing much of a slowdown in `getAccountPagesTree` on your branch vs. main.

File: packages/hash/api/src/db/postgres/link/sql/links.util.ts
Comment (luisbettencourt) [inline packages/hash/api/src/db/postgres/link/sql/links.util.ts:103]: Tried to mimick the logic for `selectAllLinksWithSourceEntity`

**Comment by luisbettencourt** [comment]
> This is really cool, Luís!
> 
> Just wanted to write down the idea that I had when we talked offline, I'm not sure it would help much after looking at the code. But here is the general outline:
> 
> What I was thinking is that given the following page structure:
> 
> Screenshot
> Which has the following ordering current (identation means a parent relation):
> 
> Current page properties
> we could instead have the indexing happen across all pages, regardless of the the parent relations, such that we get the following indexing:
> 
> Proposed page properties
> because the parent relation does not influence the order of the pages.
> 
> If I were to move `"End page"` (index `"Zz"`) above `"Middle page"` (index `"a3"`), the index would change on `"End page"` but not necessarily remove its parent page relationship.
> 
> I would imagine this would simplify some of the ordering logic a little, but from my local testing of the current code, I'm not seeing much of a slowdown in `getAccountPagesTree` on your branch vs. main.

Thank you for the suggestion, @thehabbos007. I think it would in fact simplify the code a fair bit, even in the frontend where we are ordering it as a flatlist. However, for support pages without an index (pages created before this PR that haven't yet been reordered) it will cause them to be misplaced outside their parent. I don't want to discard this option but I wouldn't really know how to tackle that issue without creating a migration, which doesn't look trivial at all.

**Comment by CiaranMn** [comment]
> for support pages without an index (pages created before this PR that haven't yet been reordered) it will cause them to be misplaced outside their parent.

I wouldn't worry about this – let's do the thing that is most maintainable / performant / simplest for the long term. 

We are likely to wipe the db when bringing the new API in anyway so there won't be any pages without indexes.

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts
Comment (nathggns) [inline packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts:19]: Feels to me like a type of something like this would be cleaner? 
```suggestion
export interface TreeElement {
  page: AccountPage;
  parentId: string;
  depth: number;
  index: number;
  expanded: boolean;
  expandable: boolean;
  collapsed: boolean;
  children?: TreeElement[];
}
```

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts
Comment (nathggns) [inline packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts:15]: This being stored here seems problematic to me - if you ever wanted to optimize this component it would get very difficult as you have to redo this entire tree calculation every time you open/close a page. Can you look at moving this out of the page state?

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts
Comment (nathggns) [inline packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts:50]: Don't you now have two fields – `parentId` and `parentPageEntityId` which are the same?

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts
Comment (nathggns) [inline packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts:52]: This is calculated state, is it really necessary to have it here?

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts
Comment (nathggns) [inline packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts:78]: Can you put a comment explaining what this does / is for? Probably a bit hard to follow for people not familiar with DND kit

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts
Comment (nathggns) [inline packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts:85]: Could call this something more meaningful

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx
Comment (nathggns) [inline packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx:127]: `16` here is a magic number but I think is supposed to match something in styles? Can this be a variable?

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts
Comment (nathggns) [inline packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts:98]: I think `dragDepth` should be the argument here, not `dragOffset` and `indentationWidth`, which are two variables not relevant to this function beyond calculating `dragDepth`.

**Review by nathggns — COMMENTED** [review]
Hey @luisbettencourt, this looks like a real substantial amount of work. Super well done for it!

I'm not able to review the API work or actually try it out right now, so I've got some comments on the code on the frontend, mostly to do with legibility / flow of data. I'm finding some of it a little difficult to follow at the moment & I think these changes could help. After that I'll do a more substantial review + try to get a UX review done too. 

Seriously well done though @luisbettencourt!

**Comment by luisbettencourt** [comment]
After further investigation, I realized that if we implemented @thehabbos007's suggestion of making indexes flat accross all pages in the page tree we would need to update the index of child pages when moving a parent page.

For example, given the page tree from the above comment: 

```
[
  {
    index: "a0",
    title: "Untitled",
  },
  {
    index: "a1",
    title: "My awesome page",
  },
  {
    index: "a2",
    title: "Top page",
  },
      {
        index: "a3",
        title: "Middle page",
      },
          {
            index: "a4",
            title: "Table Testing Page",
          },
      {
        // still follows the rest of the indices, but just not continuation of "a"
        index: "Zz", 
        title: "End page",
      },
]
``` 

If we were to move `Top page` to `My awesome page`'s position, we would end up with:

```
[
  {
    index: "a0",
    title: "Untitled",
  },
  {
    index: "a0V",
    title: "Top page",
  },
      {
        index: "a3",
        title: "Middle page",
      },
          {
            index: "a4",
            title: "Table Testing Page",
          },
      {
        // still follows the rest of the indices, but just not continuation of "a"
        index: "Zz", 
        title: "End page",
      },
  {
    index: "a1",
    title: "My awesome page",
  },
]
``` 

This happens because we only update the index of the page that is being moved. If we ordered all pages as a flat list by index we wouldn't have the children pages next to their parent anymore.

There's still the option to include the before and after indexes of the position the page is being moved to (in the above example `before = a0` and `after = a1`). Although I think it would be less resilient, that would simplify the backend logic as we wouldn't need to calculate the relative index from the absolute position as we are doing now.

File: packages/hash/frontend/src/components/hooks/useCreateSubPage.ts
Comment (luisbettencourt) [reply packages/hash/frontend/src/components/hooks/useCreateSubPage.ts:29]: Fetch results `data` is nullable for some reason. I didn't really write this code, I just separated the createSubpage from the createPage hook, for keeping it simpler to track loading states. Would you rather have `?` instead of the `!` ?

File: packages/hash/frontend/src/components/hooks/useCreateSubPage.ts
Comment (luisbettencourt) [reply packages/hash/frontend/src/components/hooks/useCreateSubPage.ts]: Way cleaner. It wasn't my code (just separated the hooks into 2 files) but made this change here: 89d1a3a5916dccb7fea53f1c1439c65e025db486

File: packages/hash/frontend/src/components/hooks/useAccountPages.ts
Comment (luisbettencourt) [reply packages/hash/frontend/src/components/hooks/useAccountPages.ts]: It was parentPageEntityId from `GetAccountPagesTreeQuery` that could be undefined whereas it couldn't in `AccountPage`. Made it an optional property here: 38aa9f0e0b78243f92488172a6bd2dd5ae41d0de

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx
Comment (luisbettencourt) [reply packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx]: Followed @CiaranMn's suggestion in 20b31c11ed1902944ecc4e56abd805fbdb33e0de

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx
Comment (luisbettencourt) [reply packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx]: Addressed in 564c8ece41bf89c189791cf7adb1b4a30ef35869

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx
Comment (luisbettencourt) [reply packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx]: Moved it inside useMemo in 84add747b5a84b956f8a9e0f75a58df321169ce6

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx
Comment (luisbettencourt) [reply packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx]: Done in ddbeb41ce27ec265fbf0a39910197bfd6daa8c1e

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx
Comment (luisbettencourt) [reply packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx]: Changed to useEffect and added comments in 4609c52713aec50e7f7cf79cb1493aec393f164d

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx
Comment (luisbettencourt) [reply packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx]: Residual code from previous iterations where I was cloning an AccountPage array instead of a TreeElement array. Removed it in a84edd29298fe75136e917d4d1b5788b504c1c2e

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts
Comment (luisbettencourt) [reply packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts]: Addressed this in d92415cc821871d1c7242cbe76e62b88f51588ab

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/page-tree-item.tsx
Comment (luisbettencourt) [reply packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/page-tree-item.tsx]: Tried to cleanup this element props in 4902d5fe9e20b34e1f24ecbffeb6ce2471edeeb1. Let me know if you have any suggestions to further simplify it

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/page-tree-item.tsx
Comment (luisbettencourt) [reply packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/page-tree-item.tsx]: I agree. Made it more readable in d4748b7786db453b854001cac0b0abb0bba3da50

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/page-tree-item.tsx
Comment (luisbettencourt) [reply packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/page-tree-item.tsx]: Thank you, didn't realize that. Added the argument to forwardRef in 5b5a42a2bd19c811b4afc0b374598042c801e792

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts
Comment (luisbettencourt) [reply packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts:82]: it's not so much about DND kit as it is about the Item Tree dragging solution. Added comments to the function in 2be59f165f2e986c4b13e26d41f7233f1d82a2e6

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts
Comment (luisbettencourt) [reply packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts]: Renamed to `expandedPages` in 7df9bb6ab05f1a323a46e212f1715ca736bc8694

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx
Comment (luisbettencourt) [reply packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx]: Moved it to a variable in c762b94d0bdc79fd988f6ed3b7364ed3f31e24a9

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts
Comment (luisbettencourt) [reply packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts]: Moved dragDepth calculation outside of this function in c762b94d0bdc79fd988f6ed3b7364ed3f31e24a9

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx
Comment (luisbettencourt) [reply packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx]: It didn't, it was the result of many iterations. Removed `pagesTreeList` completely in d47040a69cf467b8515ee658da22de359f1333ff

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts
Comment (luisbettencourt) [reply packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts]: This function was creating the pageList recursively by starting in the pages with  `depth = 0` (with no parentPage, hence the `null` there). Simplified this function a lot in 4a90efaaa7d2f35b3945367350cc54e823bf1420

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts
Comment (luisbettencourt) [reply packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts]: You're right, removed `parentId` in 676a69746a75a38c623f191453b4160ca5538b0b

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts
Comment (luisbettencourt) [reply packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts]: Followed this suggestion in 3366f44744c15d79b38e7446fe7cbbe42bf39b13

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts
Comment (luisbettencourt) [reply packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts]: Moved all collapse/expand state calculations to the page item's render function in 5cab3ac1a6df55c19655ea81a335bce2e28caeac

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts
Comment (luisbettencourt) [reply packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts]: Removed in https://github.com/hashintel/hash/commit/5cab3ac1a6df55c19655ea81a335bce2e28caeac

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx
Comment (nathggns) [inline packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx:135]: This is not how to update derivied state when other state/props changes. https://beta.reactjs.org/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes 

Let me know if this is confusing and we can chat about it

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx
Comment (nathggns) [inline packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx:157]: This ensures the cursor is restored when this component unmounts, and is also clearer. 

```suggestion
  useEffect(() => {
    if (activeId) {
      document.body.style.setProperty("cursor", "grabbing");
      
      return () => {
        if (document.body.style === "grabbing") {
          document.body.style.setProperty("cursor", "");
        }
      };
    }
  }, [activeId]);
```

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx
Comment (nathggns) [inline packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx:217]: Correct me if I'm wrong, but you're doing this twice in a row?

```ts
sortedItems.filter(({ page }) => page.parentPageEntityId === parentPageEntityId)
```

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx
Comment (nathggns) [inline packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx:1]: Have you checked how this all interacts with the new page loading skeleton stuff?

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts
Comment (nathggns) [inline packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts:25]: Does this work? If not, why not? 

```suggestion
      const item: TreeItem = {
        page,
        depth,
      };
```

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx
Comment (luisbettencourt) [reply packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx]: Thank you @nathggns, I keep falling into this trap of using `useEffect`. Addressed this in f39a7ce8a6261d522bec2d935b1475e02c9ab675

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx
Comment (luisbettencourt) [reply packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx]: Done in f066c2f938e05d47d0e1f99195c9addf2293d78a

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx
Comment (luisbettencourt) [reply packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx]: Nice spot, thanks. 3b4b6ccaf15e3b400f934ec8605d0a32998adeee

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts
Comment (luisbettencourt) [reply packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/utils.ts]: It works. Changed in 28b80ac08c80c99b30c7a7bb1f68d1418db7a2d5

File: packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx
Comment (luisbettencourt) [reply packages/hash/frontend/src/shared/layout/layout-with-sidebar/account-page-list/account-page-list.tsx:1]: I had, but because I was using a `useEffect` to derive the state I had to create an `initialLoading` state variable to prevent a jump from the skeleton pages to the actual page list: fa202ed8492032f750d96672acc20b43a68f71fa

This was not a problem anymore after f39a7ce8a6261d522bec2d935b1475e02c9ab675 so I reverted to using `pagesLoading` to check what to display (skeleton or pages) in b6ddd7f91cd6fff5e272e3f9e46cd4e2bb4e8535

**Review by nathggns — DISMISSED** [review]

**Comment by nathggns** [comment]
@luisbettencourt I haven't really reviewed the API changes. You may want to wait for a review from @CiaranMn or someone else for that.

**Review by CiaranMn — APPROVED** [review]
Great work! I've tested and it works well.

**Pull request merged**
